sky/backends/backend_utils.py

```diff
@@ -699,13 +699,29 @@ def write_cluster_config(
                 remote_identity = list(profile.values())[0]
                 break
     if remote_identity != schemas.RemoteIdentityOptions.LOCAL_CREDENTIALS.value:
-        if not cloud.supports_service_account_on_remote():
-            raise exceptions.InvalidCloudConfigs(
-                'remote_identity: SERVICE_ACCOUNT is specified in '
-                f'{skypilot_config.loaded_config_path!r} for {cloud}, but it '
-                'is not supported by this cloud. Remove the config or set: '
-                '`remote_identity: LOCAL_CREDENTIALS`.')
-        excluded_clouds = [cloud]
+        if remote_identity == schemas.RemoteIdentityOptions.NO_UPLOAD.value:
+            # If NO_UPLOAD is specified, fall back to default remote identity
+            # for downstream logic but add it to excluded_clouds to skip
+            # credential file uploads.
+            remote_identity = schemas.get_default_remote_identity(
+                str(cloud).lower())
+        else:
+            if not cloud.supports_service_account_on_remote():
+                raise exceptions.InvalidCloudConfigs(
+                    'remote_identity: SERVICE_ACCOUNT is specified in '
+                    f'{skypilot_config.loaded_config_path!r} for {cloud}, but it '
+                    'is not supported by this cloud. Remove the config or set: '
+                    '`remote_identity: LOCAL_CREDENTIALS`.')
+        excluded_clouds.add(cloud)
+
+    for cloud_str, cloud_obj in cloud_registry.CLOUD_REGISTRY.items():
+        remote_identity_config = skypilot_config.get_nested(
+            (cloud_str.lower(), 'remote_identity'), None)
+        if remote_identity_config:
+            if (remote_identity_config ==
+                    schemas.RemoteIdentityOptions.NO_UPLOAD.value):
+                excluded_clouds.add(cloud_obj)
+
```

Comment:
I found the logic for the remote identities is getting a bit complicated now. Can we add a comment at the top talking about the four cases: `LOCAL_CREDENTIALS`, `SERVICE_ACCOUNT`, `custom service account name`  and `NO_UPLOAD`

Reply:
Good point - updated comments